In [1]:
%load_ext autoreload
%autoreload 2

import sys
from pathlib import Path

ROOT = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()
sys.path.insert(0, str(ROOT))

import src
import pandas as pd

/Users/gperaza/Research/informal-jobs-model/.venv/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


We load the databases generated in Notebook 01.

In [2]:
enoe = pd.read_parquet(ROOT / "outputs/enoe_workers.parquet")
od = pd.read_parquet(ROOT / "outputs/od_workers.parquet")

For standardization purposes, the attributes, their representation in each survey, and their standardized results are as follows:

* **Gender**:

  * **ENOE**: `sex`
    * (Hombre) $\to$ `H`
    * (Mujer) $\to$ `F`
  * **OD**: `sexo_nacimiento`
    * (Hombres) $\to$ `H`
    * (Mujeres) $\to$ `F`
  * **Standardized attribute:** A `genero` column is created with values `H`, `F`, or `no_especificado`.

* **Occupation**:

  * **ENOE**: `pos_ocu`

    * (Trabajadores subordinados y remunerados) $\to$ `trabajador`
    * (Empleadores) $\to$ `trabajador`
    * (Trabajadores por cuenta propia) $\to$ `independiente`
    * (Trabajadores sin pago) $\to$ `otro`
    * (No especificado) $\to$ `no_especificado`
  * **OD**: `ocupacion_raw`

    * (Empleado) $\to$ `trabajador`
    * (Persona trabajadora del hogar) $\to$ `trabajador`
    * (Trabajador del campo) $\to$ `trabajador`
    * (Profesor) $\to$ `trabajador`
    * (Patrón o empresario) $\to$ `trabajador`
    * (Trabajador independiente) $\to$ `independiente`
    * (Hogar) $\to$ `otro`
    * (Estudiante) $\to$ `otro`
    * (Jubilado o pensionado) $\to$ `otro`
    * (Desempleado) $\to$ `otro`
  * **Standardized attribute:** An `ocupacion` column is created with values `trabajador`, `independiente`, `otro`, or `no_especificado`.

* **Age**:

  * **ENOE**: `eda`
  * **OD**: `edad`
  * **Standardization for both datasets**:

    * (0-2 years) $\to$ `0_2`
    * (3-4 years) $\to$ `3_4`
    * (5 years) $\to$ `5`
    * (6-7 years) $\to$ `6_7`
    * (8-11 years) $\to$ `8_11`
    * (12-14 years) $\to$ `12_14`
    * (15-17 years) $\to$ `15_17`
    * (18-24 years) $\to$ `18_24`
    * (25-49 years) $\to$ `25_49`
    * (50-59 years) $\to$ `50_59`
    * (60-64 years) $\to$ `60_64`
    * (65 years or more) $\to$ `65_y_mas`
  * **Standardized attribute:** An `edad_num` column is created with the numerical age, and an `edad_cat` column is created with values `0_2`, `3_4`, `5`, `6_7`, `8_11`, `12_14`, `15_17`, `18_24`, `25_49`, `50_59`, `60_64`, or `65_y_mas`.

* **Education**:

  * **ENOE**: `cs_p13_1`

    * (Ninguna) $\to$ `sin_instruccion`
    * (Preescolar) $\to$ `sin_instruccion`
    * (Primaria) $\to$ `primaria_o_secundaria`
    * (Secundaria) $\to$ `primaria_o_secundaria`
    * (Preparatoria o bachillerato) $\to$ `carrera_tecnica_o_preparatoria`
    * (Normal) $\to$ `carrera_tecnica_o_preparatoria`
    * (Carrera técnica) $\to$ `carrera_tecnica_o_preparatoria`
    * (Profesional) $\to$ `licenciatura`
    * (Maestría) $\to$ `postgrado`
    * (Doctorado) $\to$ `postgrado`
    * (No sabe) $\to$ `no_especificado`
  * **OD**: `escolaridad_raw`

    * (Ninguno) $\to$ `sin_instruccion`
    * (Kinder) $\to$ `sin_instruccion`
    * (Preescolar) $\to$ `sin_instruccion`
    * (Primaria) $\to$ `primaria_o_secundaria`
    * (Secundaria) $\to$ `primaria_o_secundaria`
    * (Normal básica) $\to$ `carrera_tecnica_o_preparatoria`
    * (Preparatoria o bachillerato) $\to$ `carrera_tecnica_o_preparatoria`
    * (Carrera técnica con secundaria) $\to$ `carrera_tecnica_o_preparatoria`
    * (Carrera técnica con preparatoria) $\to$ `carrera_tecnica_o_preparatoria`
    * (Licenciatura o profesional) $\to$ `licenciatura`
    * (Maestría o doctorado) $\to$ `postgrado`
    * (No sabe) $\to$ `no_especificado`
  * **Standardized attribute:** An `escolaridad` column is created with values `sin_instruccion`, `primaria_o_secundaria`, `carrera_tecnica_o_preparatoria`, `licenciatura`, `postgrado`, or `no_especificado`.

* **Municipality**:

  * **ENOE**: `mun`

    * (Guadalajara) $\to$ `guadalajara`
    * (Zapopan) $\to$ `zapopan`
    * (San Pedro Tlaquepaque) $\to$ `tlaquepaque`
    * (Tlajomulco de Zúñiga) $\to$ `tlajomulco`
    * (Tonalá) $\to$ `tonala`
    * (El Salto) $\to$ `el_salto`
    * (Juanacatlán) $\to$ `juanacatlan`
    * (Ixtlahuacán de los Membrillos) $\to$ `ixtlahuacan_membrillos`
    * (Zapotlanejo) $\to$ `zapotlanejo`
    * (Tala) $\to$ `tala`
    * (Other municipalities) $\to$ `otro`
  * **OD**: `municipio_raw`

    * (Guadalajara) $\to$ `guadalajara`
    * (Zapopan) $\to$ `zapopan`
    * (San Pedro Tlaquepaque) $\to$ `tlaquepaque`
    * (Tlajomulco de Zúñiga) $\to$ `tlajomulco`
    * (Tonalá) $\to$ `tonala`
    * (El Salto) $\to$ `el_salto`
    * (Juanacatlán) $\to$ `juanacatlan`
    * (Ixtlahuacán de los Membrillos) $\to$ `ixtlahuacan_membrillos`
    * (Zapotlanejo) $\to$ `zapotlanejo`
    * (Tala) $\to$ `tala`
    * (Other municipalities) $\to$ `otro`
  * **Standardized attribute:** A `municipio` column is created with standardized municipality names for the municipalities of interest, while all remaining municipalities are classified as `otro`.

* **Marital status**:

  * **ENOE**: `e_con`

    * (Unión libre) $\to$ `union_libre`
    * (Separado) $\to$ `separado`
    * (Divorciado) $\to$ `divorciado`
    * (Viudo) $\to$ `viudo`
    * (Casado) $\to$ `casado`
    * (Soltero) $\to$ `soltero`
  * **OD**: `estado_civil_raw`

    * (Unión libre) $\to$ `union_libre`
    * (Separado) $\to$ `separado`
    * (Divorciado) $\to$ `divorciado`
    * (Viudo) $\to$ `viudo`
    * (Casado) $\to$ `casado`
    * (Soltero) $\to$ `soltero`
  * **Standardized attribute:** An `estado_civil` column is created with values `union_libre`, `separado`, `divorciado`, `viudo`, `casado`, `soltero`, or `no_especificado`.

* **Relationship to household head**:

  * **ENOE**: `par_c`

    * (Head of household) $\to$ `jefe_del_hogar`
    * (Spouse or partner) $\to$ `conyuge`
    * (Son or daughter) $\to$ `hijo`
    * (Other relative) $\to$ `otro_parentesco`
    * (Non-relative) $\to$ `sin_parentesco`
  * **OD**: `parentesco_raw`

    * (Jefe del hogar) $\to$ `jefe_del_hogar`
    * (Cónyuge) $\to$ `conyuge`
    * (Compañero) $\to$ `conyuge`
    * (Hijo) $\to$ `hijo`
    * (Nieto) $\to$ `otro_parentesco`
    * (Otro parentesco) $\to$ `otro_parentesco`
    * (Sin parentesco) $\to$ `sin_parentesco`
  * **Standardized attribute:** A `parentesco` column is created with values `jefe_del_hogar`, `conyuge`, `hijo`, `otro_parentesco`, `sin_parentesco`, or `no_especificado`.

* **Dwelling size**:

  * **ENOE**: `dwelling_size`

    * (1 person) $\to$ `1`
    * (2 people) $\to$ `2`
    * (3 people) $\to$ `3`
    * $\dots$
    * (9 people) $\to$ `9`
    * (10 or more people) $\to$ `10_y_mas`
  * **OD**: `dwelling_size`

    * (1 person) $\to$ `1`
    * (2 people) $\to$ `2`
    * (3 people) $\to$ `3`
    * $\dots$
    * (9 people) $\to$ `9`
    * (10 y + / 10 y más) $\to$ `10_y_mas`
  * **Standardized attribute:** A `tamano_viv_num` column is created with the numerical dwelling size, and a `tamano_viv_cat` column is created with categories from `1` to `9` and `10_y_mas`.

* **Economic sector**:

  * **ENOE**: `scian`

    * (Agricultura, ganadería, aprovechamiento forestal, pesca y caza) $\to$ `gobierno_otro_agricultura`
    * (Minería) $\to$ `gobierno_otro_agricultura`
    * (Electricidad, agua y suministro de gas) $\to$ `gobierno_otro_agricultura`
    * (Construcción) $\to$ `manufactura_construccion`
    * (Industrias manufactureras) $\to$ `manufactura_construccion`
    * (Comercio al por mayor) $\to$ `comercio`
    * (Comercio al por menor) $\to$ `comercio`
    * (Transportes, correos y almacenamiento) $\to$ `servicios_transporte`
    * (Información en medios masivos) $\to$ `gobierno_otro_agricultura`
    * (Servicios financieros y de seguros) $\to$ `servicios_transporte`
    * (Servicios inmobiliarios y de alquiler) $\to$ `servicios_transporte`
    * (Servicios profesionales, científicos y técnicos) $\to$ `servicios_transporte`
    * (Corporativos) $\to$ `servicios_transporte`
    * (Servicios de apoyo a los negocios y manejo de desechos) $\to$ `servicios_transporte`
    * (Servicios educativos) $\to$ `servicios_transporte`
    * (Servicios de salud y asistencia social) $\to$ `servicios_transporte`
    * (Servicios de esparcimiento, culturales y deportivos) $\to$ `servicios_transporte`
    * (Servicios de alojamiento temporal y preparación de alimentos y bebidas) $\to$ `servicios_transporte`
    * (Otros servicios) $\to$ `servicios_transporte`
    * (Gobierno y organismos internacionales) $\to$ `gobierno_otro_agricultura`
    * (No especificado) $\to$ `no_especificado`
  * **OD**: `giro_empresa`

    * (Comercio) $\to$ `comercio`
    * (Servicio) $\to$ `servicios_transporte`
    * (Educación) $\to$ `servicios_transporte`
    * (Industria) $\to$ `manufactura_construccion`
    * (Gobierno/sector público) $\to$ `gobierno_otro_agricultura`
  * **Standardized attribute:** A `sector` column is created with values `comercio`, `servicios_transporte`, `manufactura_construccion`, `gobierno_otro_agricultura`, or `no_especificado`. A `sector_desconocido` column is also created to identify observations whose sector cannot be determined.

* **Informality**:

  * **ENOE**: `emp_ppal`

    * (Empleo informal) $\to$ `1`
    * (Empleo formal) $\to$ `0`
  * **OD**: Not available
  * **Standardized attribute:** An `informal` column is created only for ENOE, where `1` indicates an informal worker and `0` indicates a formal worker.


In [3]:
enoe_harmonized = src.harmonize_enoe_dataframe(enoe)
od_harmonized = src.harmonize_od_dataframe(od)

In [4]:
common_columns = ["genero", "ocupacion", "edad_num", "edad_cat", "escolaridad", "municipio", "estado_civil", "parentesco", "tamano_viv_num", "tamano_viv_cat", "sector", "sector_desconocido"]

assert enoe_harmonized.shape[0] == enoe.shape[0], "ENOE row count changed during harmonization."
assert od_harmonized.shape[0] == od.shape[0], "OD row count changed during harmonization."
assert set(common_columns).issubset(enoe_harmonized.columns), "Missing harmonized columns in ENOE."
assert set(common_columns).issubset(od_harmonized.columns), "Missing harmonized columns in OD."
assert "informal" in enoe_harmonized.columns, "Missing informal label in ENOE."

print("Harmonization validation completed successfully.")

Harmonization validation completed successfully.


In [5]:
for column in common_columns:
    print(f"\n{column}")
    print("ENOE:", enoe_harmonized[column].value_counts(dropna=False).to_dict())
    print("OD:", od_harmonized[column].value_counts(dropna=False).to_dict())


genero
ENOE: {'H': 3959, 'F': 2834}
OD: {'H': 16684, 'F': 10229}

ocupacion
ENOE: {'trabajador': 5391, 'independiente': 1235, 'otro': 167}
OD: {'trabajador': 23274, 'independiente': 3417, 'otro': 194, 'no_especificado': 28}

edad_num
ENOE: {np.int64(27): 190, np.int64(24): 189, np.int64(26): 186, np.int64(29): 176, np.int64(22): 176, np.int64(23): 175, np.int64(50): 175, np.int64(31): 173, np.int64(30): 167, np.int64(32): 167, np.int64(28): 165, np.int64(33): 165, np.int64(38): 160, np.int64(36): 155, np.int64(35): 153, np.int64(40): 153, np.int64(37): 152, np.int64(41): 146, np.int64(46): 145, np.int64(39): 144, np.int64(34): 143, np.int64(45): 142, np.int64(20): 139, np.int64(42): 139, np.int64(53): 138, np.int64(43): 138, np.int64(21): 135, np.int64(44): 135, np.int64(49): 135, np.int64(25): 134, np.int64(54): 128, np.int64(51): 127, np.int64(52): 127, np.int64(48): 125, np.int64(19): 113, np.int64(47): 110, np.int64(55): 109, np.int64(18): 108, np.int64(56): 102, np.int64(57): 87,

Since we are going to some classifications in the pipeline, it is important to analyze the distribution of those attributes with unspecified values or marked as “other.”

In [6]:
print("ENOE municipalities mapped as 'otro':")
print(enoe_harmonized.loc[enoe_harmonized["municipio"] == "otro", "mun"].value_counts(dropna=False))

print("\nOD municipalities mapped as 'otro':")
print(od_harmonized.loc[od_harmonized["municipio"] == "otro", "municipio_raw"].value_counts(dropna=False))

print("OD education mapped as 'no_especificado':")
print(od_harmonized.loc[od_harmonized["escolaridad"] == "no_especificado", "escolaridad_raw"].value_counts(dropna=False))

print("OD sectors mapped as 'no_especificado':")
print(od_harmonized.loc[od_harmonized["sector_desconocido"], "giro_empresa"].value_counts(dropna=False))

ENOE municipalities mapped as 'otro':
mun
23     148
123    142
93     120
67     118
18     101
82      95
6       80
53      62
37      60
50      55
35      53
80      51
100     43
36      39
81      36
73      36
10      36
22      33
47      33
63      32
1       32
66      32
13      27
108     27
45      25
119     24
74      24
27      22
68      17
58      15
30      13
Name: count, dtype: Int64

OD municipalities mapped as 'otro':
Series([], Name: count, dtype: int64[pyarrow])
OD education mapped as 'no_especificado':
escolaridad_raw
<NA>       5037
No sabe     158
Name: count, dtype: int64[pyarrow]
OD sectors mapped as 'no_especificado':
giro_empresa
<NA>    9484
Name: count, dtype: int64[pyarrow]


In this distribution, we can see that, for ENOE, there is a considerable number of observations whose municipalities are classified as `other`, since they are not part of the set of municipalities approved by the OD. For this reason, the comparative and classification analysis between the two databases will be conducted by considering only the geography common to both sources.

Furthermore, in the education variable, there are 5,195 observations from the OD classified as `unspecified`. This result will be relevant during the construction of the machine learning models, as it will allow us to evaluate the contribution of educational attainment to model performance and to define how to handle observations for which this information is not available.

Finally, the 9,484 observations from the OD whose economic sector was classified as `no_specified` correspond to workers whose sector must be imputed later using the machine learning model, assigning them to one of the previously validated economic sectors.

In [7]:
print(f"ENOE workers: {len(enoe_harmonized):,}")
print(f"OD workers: {len(od_harmonized):,}")
print(f"ENOE unknown sectors: {enoe_harmonized['sector_desconocido'].sum():,}")
print(f"OD unknown sectors: {od_harmonized['sector_desconocido'].sum():,}")

ENOE workers: 6,793
OD workers: 26,913
ENOE unknown sectors: 37
OD unknown sectors: 9,484


In [8]:
output_directory = ROOT / "outputs"
output_directory.mkdir(exist_ok=True)

enoe_harmonized.to_parquet(output_directory / "enoe_harmonized.parquet", index=False)
od_harmonized.to_parquet(output_directory / "od_harmonized.parquet", index=False)